Scalar.Common/Maintenance/ConfigStep.cs

```diff
@@ -208,25 +209,86 @@ public bool TrySetConfig(out string error)
                 return false;
             }
 
-            return this.ConfigureWatchmanIntegration(out error);
+            this.GetConfigSettings();
+
+            if (this.existingConfigSettings == null)
+            {
+                return this.ConfigureWatchmanIntegration(out error);
+            }
+
+            GitProcess.ConfigResult config = null;
+            GitProcess.Result getResult = this.RunGitCommand(
+                process => {
+                    config = process.GetFromLocalConfig("feature.scalar");
+                    return null;
+                },
+                nameof(GitProcess.GetFromLocalConfig)
+            );
+            GitFeatureFlags flags = GitVersion.GetAvailableGitFeatures(this.Context.Tracer);
+            config.TryParseAsString(out string scalar, out error, defaultValue: "true");
+
+            if (scalar.Equals("false"))
+            {
+                GitProcess.Result deleteResult = this.RunGitCommand(
+                    process => process.DeleteFromLocalConfig("core.fsmonitor"),
+                    nameof(GitProcess.DeleteFromLocalConfig)
+                );
+
+                return deleteResult.ExitCodeIsSuccess;
+            }
+            else if (scalar.Equals("experimental")
+                     // Make sure Git supports builtin FS Monitor
+                     && flags.HasFlag(GitFeatureFlags.BuiltinFSMonitor)
+                     // For now, this doesn't work on Linux
+                     && !RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                // ":internal:" is a custom value to specify the builtin
+                // FS Monitor feature.
+                GitProcess.Result setResult = this.RunGitCommand(
+                    process => process.SetInLocalConfig("core.fsmonitor", ":internal:"),
+                    nameof(GitProcess.SetInLocalConfig)
+                );
+
+                return setResult.ExitCodeIsSuccess;
+            }
+            else
+            {
+                return this.ConfigureWatchmanIntegration(out error);
+            }
         }
 
         protected override void PerformMaintenance()
         {
             this.TrySetConfig(out _);
         }
 
-        private bool TrySetConfig(Dictionary<string, string> configSettings, bool isRequired, out string error, bool add = false)
+        private Dictionary<string, GitConfigSetting> GetConfigSettings()
         {
-            Dictionary<string, GitConfigSetting> existingConfigSettings = null;
+            if (this.existingConfigSettings != null)
+            {
+                return this.existingConfigSettings;
+            }
 
             GitProcess.Result getConfigResult = this.RunGitCommand(
-                process => process.TryGetAllConfig(localOnly: isRequired, configSettings: out existingConfigSettings),
+                process => process.TryGetAllConfig(localOnly: true, configSettings: out this.existingConfigSettings),
```

Comment:
(!) We're setting both a field `this.existingConfigSettings` _and_ returning the result? (but not actually using the returned result, and instead relying on the method setting this field as a side effect above `TrySetConfig`).

Either don't set the field and always use the returned value, or change this to a `void` method that's named `EnsureConfigSettingsSet` or something that makes it clear the field is being mutated.

Reply:
I've modified it to return the value and only mutate it to non-null. All callers use the result.